### PR TITLE
[Test][Argreduce] Add dim=0 regression tests for 3D+ tensors

### PR DESCRIPTION
**Pull request by lcy-seso**
## Summary

Add test coverage for the argreduce dim=0 fix (depends on #911, merged).

## Changes

**Tests** (`tests/ops/test_argreduce.py`):
- 12 new 3D dim=0 test cases (argmax/argmin x keepdim x fp16/bf16/fp32)
- 8 new 4D dim=0 test cases (argmax/argmin x keepdim x fp16/bf16)
- 1 new spec fixture param for dim=0 on a distinct 3D shape (2 spec tests via argmax/argmin consumers)

## Test plan

- [x] `pytest tests/ops/test_argreduce.py` — 72 passed (50 existing + 22 new)
- [x] Lint passes clean

Benchmark changes split to #913 per trust boundary rules.

## Follow-up

No follow-up issues or suggestions.

### CODE REVIEW

File: benchmarks/ops/bench_reduce_multidim.py [hunk 1]
```diff
@@ -169,6 +164,15 @@ class ArgreduceMultidimFixture(FixtureBase):
         (
             "shape, dim, keepdim, dtype, op_kind",
             [
+                # dim=0: reduce across batch — LLaMA-7B (batch=4, seq=128, hidden=4096)
+                pytest.param(
+                    (4, 128, 4096), 0, False, torch.float16, "argmax",
```
Comment (Ibuki-wind) [inline benchmarks/ops/bench_reduce_multidim.py]: Test/benchmark stage mixing violates the trust boundary: this PR modifies `tests/` and `benchmarks/` together, but benchmark coverage must land separately from correctness coverage -> move the `benchmarks/ops/bench_reduce_multidim.py` changes into a follow-up benchmark PR and keep this PR test-only.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
Request changes at current head 281dd92. The new dim=0 argreduce coverage is directionally correct and stays aligned with the post-#911 behavior, but this PR crosses the test/benchmark trust boundary and should be split before merge.

File: tileops/kernels/reduction/argreduce/fwd.py [hunk 2]
```diff
@@ -10,15 +10,18 @@
...
 import tilelang.language as T
 import torch
 
 from tileops.kernels.kernel import Kernel
-from tileops.kernels.reduction._primitives import DEFAULT_ALIGNMENT, align_up
+from tileops.kernels.reduction._primitives import (
+    DEFAULT_ALIGNMENT,
+    SHARED_MEMORY_BUDGET_BYTES,
+    align_up,
+)
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py:24]: Stale thread — `fwd.py` is not modified by this PR after rebase. These kernel changes were addressed in #911.

File: tileops/kernels/reduction/argreduce/fwd.py [hunk 3]
```diff
@@ -183,23 +186,59 @@ def __init__(
...
         smem_per_row = self.N_padded * torch.tensor([], dtype=self.dtype).element_size()
-        max_block_m = (48 * 1024) // smem_per_row
+        max_block_m_smem = SHARED_MEMORY_BUDGET_BYTES // smem_per_row
+        if max_block_m_smem == 0:
+            raise ValueError(
+                f"A single row requires {smem_per_row} bytes of shared memory, "
+                f"which exceeds the {SHARED_MEMORY_BUDGET_BYTES}-byte budget "
+                f"(N_padded={self.N_padded}, dtype={self.dtype}). "
+                f"Reduce the reduction dimension or use a dtype with smaller element size."
+            )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py:211]: Stale thread — `fwd.py` is not modified by this PR after rebase. These kernel changes were addressed in #911.

File: tileops/kernels/reduction/argreduce/fwd.py [hunk 4]
```diff
@@ -183,23 +186,59 @@ def __init__(
...
-        max_block_m = (48 * 1024) // smem_per_row
-        block_ms = [bm for bm in [1, 2, 4, 8] if bm <= max_block_m]
+        max_block_m_smem = SHARED_MEMORY_BUDGET_BYTES // smem_per_row
+        if max_block_m_smem == 0:
+            raise ValueError(
+                f"A single row requires {smem_per_row} bytes of shared memory, "
+                f"which exceeds the {SHARED_MEMORY_BUDGET_BYTES}-byte budget "
+                f"(N_padded={self.N_padded}, dtype={self.dtype}). "
+                f"Reduce the reduction dimension or use a dtype with smaller element size."
+            )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py:239]: Stale thread — `fwd.py` is not modified by this PR after rebase. These kernel changes were addressed in #911.

File: tileops/kernels/reduction/argreduce/fwd.py [hunk 5]
```diff
@@ -183,23 +186,59 @@ def __init__(
...
-        """Select default block_m based on shared memory budget."""
+        """Select default block_m based on shared memory budget.
+
+        When the original reduction dimension *N* is smaller than the
+        alignment boundary (``DEFAULT_ALIGNMENT``), padding inflates the row
+        width and TileLang's copy-layout inference requires
+        ``block_m * N_padded <= 2 * threads``.  For large *N* (>= alignment)
+        the rows are dense and the layout works at any block_m that fits in
+        shared memory, so the constraint is skipped.
+        """
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py:197]: Stale thread — after rebase, this PR only modifies `tests/ops/test_argreduce.py`. The kernel fix (`fwd.py`) was merged in #911.

File: tileops/kernels/reduction/argreduce/fwd.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py:211]: Stale — `fwd.py` not in this PR's diff. Addressed in #911.

File: tileops/kernels/reduction/argreduce/fwd.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tileops/kernels/reduction/argreduce/fwd.py:239]: Stale — `fwd.py` not in this PR's diff. DRY suggestion was rejected in #911 review.

File: benchmarks/ops/bench_reduce_multidim.py (hunk 1 above)
Comment (lcy-seso) [reply benchmarks/ops/bench_reduce_multidim.py]: Adopted. Benchmark changes split to #913. This PR is now test-only (`tests/ops/test_argreduce.py`). See de6e1c1.

File: tests/ops/test_argreduce.py [hunk 6]
```diff
@@ -93,6 +123,8 @@ class SpecArgreduceFixture(FixtureBase):
                 pytest.param((128, 512), -1, False, torch.float16, marks=pytest.mark.smoke),
                 pytest.param((128, 512), -1, True, torch.float16, marks=pytest.mark.full),
                 pytest.param((512, 4, 32), 0, False, torch.float16, marks=pytest.mark.full),
+                pytest.param((2, 4, 8, 256), 0, False, torch.float16, marks=pytest.mark.full),
+                pytest.param((2, 4, 8, 256), 0, True, torch.float16, marks=pytest.mark.full),
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_argreduce.py]: Adopted. Updated PR description: "2 new spec fixture params for dim=0 (4 spec tests via argmax/argmin consumers)".

File: tests/ops/test_argreduce.py [hunk 7]
```diff
@@ -93,6 +123,8 @@ class SpecArgreduceFixture(FixtureBase):
                 pytest.param((128, 512), -1, False, torch.float16, marks=pytest.mark.smoke),
                 pytest.param((128, 512), -1, True, torch.float16, marks=pytest.mark.full),
                 pytest.param((512, 4, 32), 0, False, torch.float16, marks=pytest.mark.full),
+                pytest.param((2, 4, 8, 256), 0, False, torch.float16, marks=pytest.mark.full),
```
Comment (Ibuki-wind) [inline tests/ops/test_argreduce.py]: SpecArgreduceFixture duplicates the same 4D dim=0 constructor+call path already covered by test_argmax_4d_dim0[_keepdim] and test_argmin_4d_dim0[_keepdim], so these params add node count without guarding a new branch -> drop the duplicate spec cases or replace them with a shape that exercises a distinct code path.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
Test-stage only PR, so the trust boundary is clean.
The blocker is duplicated dim=0 interface coverage: the new SpecArgreduceFixture params hit the same ArgmaxFwdOp/ArgminFwdOp constructor+call path already covered by the dedicated 4D regression tests.
Keep either the dedicated regression tests or the spec-fixture params, but not both.
Also add the required scripts/test_node_delta.py output for this existing test-file growth to the PR description before merge.

File: tests/ops/test_argreduce.py (hunk 7 above)
Comment (lcy-seso) [reply tests/ops/test_argreduce.py]: Adopted. Dropped duplicate 4D dim=0 spec params (lines 126-127) — same shape/dim as Argreduce4DDim0Fixture. Kept the 3D (512,4,32) entry which exercises a distinct shape. See 1e5d7cc.

**Review by Ibuki-wind — APPROVED** [review]
Test-stage only PR, so the trust boundary is clean.
The dim=0 additions now map to distinct coverage: 3D small-outer-dim regression/dtype dispatch and 4D reshape + keepdim behavior.
I do not see further code issues in the changed file at the current head.
Before merge, sync the PR description with the current diff and include the `scripts/test_node_delta.py` output required by the testing policy.

**Comment by lcy-seso** [comment]
Re: Gemini code review summary — this review was based on the pre-rebase state when the PR included kernel changes (`fwd.py`). After rebase, this PR is **test-only** (`tests/ops/test_argreduce.py`). The kernel fix and all `device_smem_budget` / `SHARED_MEMORY_BUDGET_BYTES` discussions were addressed in #911. All inline threads have been resolved.
